### PR TITLE
Option to configure default analysis types in SegmentMetadataQuery

### PR DESCRIPTION
**Pull request by gkc2104**
https://github.com/druid-io/druid/pull/3773

Now that `defaultAnalysisTypes` is being updated, it would make it easier for users to update to `0.10.0` if they could easily revert to old behavior without having to update all SegmentMetadataQueries they use.

Suggest not only providing an option to revert to old behavior but also be able to specify their own defaultAnalysisTypes using  `druid.query.segmentMetadata.defaultAnalysisType`.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryConfig.java [hunk 1]
```diff
@@ -20,19 +20,29 @@
...
+      SegmentMetadataQuery.AnalysisType.CARDINALITY,
+      SegmentMetadataQuery.AnalysisType.INTERVAL,
+      SegmentMetadataQuery.AnalysisType.MINMAX
+  );
 
   @JsonProperty
   private Period defaultHistory = ISO_FORMATTER.parsePeriod(DEFAULT_PERIOD_STRING);
 
+  @JsonProperty
+  private EnumSet<SegmentMetadataQuery.AnalysisType> defaultAnalysisType = DEFAULT_ANALYSIS_TYPES;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryConfig.java]: Suggested defaultAnalysisType**s** (update getter and setter names too)

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 2]
```diff
@@ -404,4 +405,39 @@ public static SegmentAnalysis finalizeAnalysis(SegmentAnalysis analysis)
         analysis.isRollup()
     );
   }
+
+  public EnumSet<SegmentMetadataQuery.AnalysisType> getAnalysisTypes(SegmentMetadataQuery query)
+  {
+    if (query.getAnalysisTypes() == null) {
+      return config != null ? config.getDefaultAnalysisTypes() : SegmentMetadataQueryConfig.DEFAULT_ANALYSIS_TYPES;
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: It seems like handling null config shouldn't be necessary here. It shouldn't be null in production. If it is null sometimes in testing, how about replacing those spots with `new SegmentMetadataQueryConfig()` to get a default config? Then, `SegmentMetadataQueryConfig.DEFAULT_ANALYSIS_TYPES` could be private.

**Comment by gkc2104** [comment]
Given that the config is specific to the broker's `runtime.properties`, would other nodes like historical's be able to have their own `SegmentMetadataQueryConfig`, as this where it gets set, and that would also mean that they would be using different ToolChests as well, is it intended for them to be able to set different `defaultHistory` as well.

It has been reimplemented to not be reliant on the query for analysisTypes in `SegmentMetadataQueryQueryToolChest `, I extract it from the ToolChest in
https://github.com/metamx/druid/blob/defaultAnalysisUpdate/processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryRunnerFactory.java#L93
and use that at all the places analysisTypes is required in the QueryRunnerFactory, to rely on analysisTypes of the config, if not in the query.

File: docs/content/configuration/broker.md [hunk 3]
```diff
@@ -86,6 +86,7 @@ See [groupBy server configuration](../querying/groupbyquery.html#server-configur
 |Property|Description|Default|
 |--------|-----------|-------|
 |`druid.query.segmentMetadata.defaultHistory`|When no interval is specified in the query, use a default interval of defaultHistory before the end time of the most recent segment, specified in ISO8601 format. This property also controls the duration of the default interval used by GET /druid/v2/datasources/{dataSourceName} interactions for retrieving datasource dimensions/metrics.|P1W|
+|`druid.query.segmentMetadata.defaultAnalysisTypes`|This can be used to set the Default Analysis Types for all segment metadata queries, this can be overridden when making the query|[CARDINALITY, INTERVAL, MINMAX]|
```
Comment (leventov) [inline docs/content/configuration/broker.md]: The default value here should be updated to `["cardinality", "interval", "minmax"]`. Also please point to this config in docs of Segment Metadata Queries, line 35.

**Comment by leventov** [comment]
@fjy this PR actually blocks us from updating to **0.10.0** (even not 0.10.1), so it fits the criteria that I suggested before, unlike many other PRs that currently target 0.10.1, and don't block anybody from updating to 0.10.0.

**Comment by drcrallen** [comment]
I think @gianm is proposing that the broker overwrite the analysis types if null is passed in, and if someone sets `druid.query.segmentMetadata.defaultAnalysisType` differently on historicals than brokers, then the broker setting should be enforced. If someone queries historicals directly, then caveat emptor.

**Comment by gianm** [comment]
Yeah, I meant what @drcrallen said.

In particular I'm suggesting that if the broker receives a query without analysisTypes set, it should rewrite the query to include analysisTypes before it passes it down to other nodes. Otherwise different nodes involved in the query might not agree on what the analysisTypes are, which could cause weird results.

**Comment by gkc2104** [comment]
I'm not sure if this was the best way of doing it, but I made the `analysisType` field of `SegmentMetadatQuery` not final, and provided a setter, that is used to updated the query in all calls of query in the overridden methods of `QueryToolChest` not only `mergeResults`, and also defaulted the cache strategy and other methods that relied on analysis Tupe to be provided by the query

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 4]
```diff
@@ -166,12 +169,66 @@ public String getType()
     return analysisTypes;
   }
 
+  public void setAnalysisTypes(EnumSet<AnalysisType> analysisTypes)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Query object should be immutable. Please replace this method with "withAnalysisTypes()", returning a new query object with the given analysis types. Also update SegmentMetadataQueryBuilder, to include a new field.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 5]
```diff
@@ -107,12 +108,14 @@ public SegmentMetadataQueryQueryToolChest(SegmentMetadataQueryConfig config, Gen
           Map<String, Object> context
       )
       {
-        Query<SegmentAnalysis> query = queryPlus.getQuery();
+        SegmentMetadataQuery castedQuery = (SegmentMetadataQuery) queryPlus.getQuery();
+        SegmentMetadataQuery updatedQuery =(SegmentMetadataQuery) (castedQuery.withAnalysisTypes(getFinalAnalysisTypes(castedQuery)));
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: `withAnalysisTypes()` could return `SegmentMetadataQuery` and casting not needed.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 6]
```diff
@@ -406,4 +418,14 @@ public static SegmentAnalysis finalizeAnalysis(SegmentAnalysis analysis)
         analysis.isRollup()
     );
   }
+
+  public EnumSet<SegmentMetadataQuery.AnalysisType> getFinalAnalysisTypes(SegmentMetadataQuery query)
+  {
+    if (query.getAnalysisTypes() == null) {
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Suggested `Objects.firstNonNull()`

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 5 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Suggested to join those methods into a single method like `SegmentMetadataQuery.withFinalizedAnalysisTypes(toolChest)`, for less boilerplate, because currently they are always used together. Also this method may avoid creating a new query object and return the same instance, if analysisTypes are already set.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 7]
```diff
@@ -135,15 +141,17 @@ public int compare(
...
 
       @Override
       protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final Query<SegmentAnalysis> inQ)
       {
+        SegmentMetadataQuery query = (SegmentMetadataQuery) inQ;
+        SegmentMetadataQuery updatedQuery = (SegmentMetadataQuery) query.withAnalysisTypes(getFinalAnalysisTypes(query));
         return new BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis>()
         {
-          private final SegmentMetadataQuery query = (SegmentMetadataQuery) inQ;
+          private final SegmentMetadataQuery query = updatedQuery;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Why this field is needed?

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 8]
```diff
@@ -121,7 +124,10 @@ public SegmentMetadataQueryQueryToolChest(SegmentMetadataQueryConfig config, Gen
       @Override
       protected Ordering<SegmentAnalysis> makeOrdering(Query<SegmentAnalysis> query)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: There seem to be no point for this anonymous QueryRunner to extend `ResultMergeQueryRunner` rather than `BySegmentSkippingQueryRunner` directly, because it overrides `doRun()`. Then this QueryRunner doesn't need to follow API of `ResultMergeQueryRunner`, it may accept `SegmentMetadataQuery` instead of `Query<SegmentAnalysis>` in `makeOrdering()` and `createMergeFn()`, to avoid casting.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 9]
```diff
@@ -121,7 +124,10 @@ public SegmentMetadataQueryQueryToolChest(SegmentMetadataQueryConfig config, Gen
...
       {
-        if (((SegmentMetadataQuery) query).isMerge()) {
+        SegmentMetadataQuery castedQuery = (SegmentMetadataQuery) query;
+        SegmentMetadataQuery updatedQuery =(SegmentMetadataQuery) (castedQuery.withAnalysisTypes(getFinalAnalysisTypes(castedQuery)));
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: This is already done in upstream doRun() call.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 10]
```diff
@@ -135,15 +141,17 @@ public int compare(
...
 
-        return query.getResultOrdering(); // No two elements should be equal, so it should never merge
+        return updatedQuery.getResultOrdering(); // No two elements should be equal, so it should never merge
       }
 
       @Override
       protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final Query<SegmentAnalysis> inQ)
       {
+        SegmentMetadataQuery query = (SegmentMetadataQuery) inQ;
+        SegmentMetadataQuery updatedQuery = (SegmentMetadataQuery) query.withAnalysisTypes(getFinalAnalysisTypes(query));
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: This is already done in upstream doRun() call.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 11]
```diff
@@ -158,6 +166,7 @@ public SegmentAnalysis apply(SegmentAnalysis arg1, SegmentAnalysis arg2)
   @Override
   public QueryMetrics<Query<?>> makeMetrics(SegmentMetadataQuery query)
   {
+    query.withAnalysisTypes(getFinalAnalysisTypes(query));
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: The result of this call is unused. I don't think it should be here.

File: processing/src/main/java/io/druid/query/Druids.java [hunk 12]
```diff
@@ -1054,6 +1057,13 @@ public SegmentMetadataQueryBuilder analysisTypes(EnumSet<SegmentMetadataQuery.An
       return this;
     }
 
+    public SegmentMetadataQueryBuilder usingDefaultIntervals(Boolean usingDefaultInterval)
+    {
+      this.usingDefaultInterval = usingDefaultInterval;
+      return this;
+    }
+
+
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/Druids.java]: Extra line

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 13]
```diff
@@ -178,19 +187,21 @@ public SegmentAnalysis apply(SegmentAnalysis arg1, SegmentAnalysis arg2)
   @Override
   public CacheStrategy<SegmentAnalysis, SegmentAnalysis, SegmentMetadataQuery> getCacheStrategy(final SegmentMetadataQuery query)
   {
+
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Extra line

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 14]
```diff
@@ -178,19 +187,21 @@ public SegmentAnalysis apply(SegmentAnalysis arg1, SegmentAnalysis arg2)
...
     return new CacheStrategy<SegmentAnalysis, SegmentAnalysis, SegmentMetadataQuery>()
     {
       @Override
-      public boolean isCacheable(SegmentMetadataQuery query, boolean willMergeRunners)
+      public boolean isCacheable(SegmentMetadataQuery updatedQuery, boolean willMergeRunners)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Shouldn't be renamed?

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 15]
```diff
@@ -235,7 +246,8 @@ public SegmentAnalysis apply(@Nullable SegmentAnalysis input)
   @Override
   public <T extends LogicalSegment> List<T> filterSegments(SegmentMetadataQuery query, List<T> segments)
   {
-    if (!query.isUsingDefaultInterval()) {
+    SegmentMetadataQuery updatedQuery = (SegmentMetadataQuery) query.withAnalysisTypes(getFinalAnalysisTypes(query));
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: This is pointless, because doesn't affect isUsingDefaultInterval

File: processing/src/main/java/io/druid/query/Druids.java [hunk 16]
```diff
@@ -962,6 +962,7 @@ public static ResultBuilder newResultBuilder()
     private Boolean merge;
     private Boolean lenientAggregatorMerge;
     private Map<String, Object> context;
+    private Boolean usingDefaultInterval;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/Druids.java]: useDefaultInterval seems to be an unnecessary configuration, that allows to create inconsistency, if you pass useDefaultInterval=false, and querySegmentSpec which actually represents the default interval to SegmentMetadataQuery constructor.

I suggest the following plan:
 - Don't add usingDefaultInterval to this builder
 - Leave usingDefaultInterval parameter of SegmentMetadataQuery for compatibility, but ignore it, and document the fact that it is going to be removed.
 - In the constructor, set usingDefaultInterval=true if querySegmentSpec == null or querySegmentSpec is not null, and it has just one interval which is equal to the default interval.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 17]
```diff
@@ -138,12 +138,11 @@ public int compare(
...
       }
 
-      @Override
-      protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final Query<SegmentAnalysis> inQ)
+      protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final SegmentMetadataQuery inQ)
       {
         return new BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis>()
         {
-          private final SegmentMetadataQuery query = (SegmentMetadataQuery) inQ;
+          private final SegmentMetadataQuery query = inQ;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: This field is not needed

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 18]
```diff
@@ -107,21 +107,21 @@ public SegmentMetadataQueryQueryToolChest(SegmentMetadataQueryConfig config, Gen
...
+                makeOrdering(updatedQuery),
+                createMergeFn(updatedQuery)
             ),
             MERGE_TRANSFORM_FN
         );
       }
 
-      @Override
-      protected Ordering<SegmentAnalysis> makeOrdering(Query<SegmentAnalysis> query)
+      protected Ordering<SegmentAnalysis> makeOrdering(SegmentMetadataQuery query)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Could be `private`

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 19]
```diff
@@ -138,12 +138,11 @@ public int compare(
         return query.getResultOrdering(); // No two elements should be equal, so it should never merge
       }
 
-      @Override
-      protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final Query<SegmentAnalysis> inQ)
+      protected BinaryFn<SegmentAnalysis, SegmentAnalysis, SegmentAnalysis> createMergeFn(final SegmentMetadataQuery inQ)
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Could be `private`

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 20]
```diff
@@ -32,8 +32,10 @@
 import io.druid.query.TableDataSource;
 import io.druid.query.UnionDataSource;
 import io.druid.query.filter.DimFilter;
+import io.druid.query.metadata.SegmentMetadataQueryConfig;
 import io.druid.query.spec.MultipleIntervalSegmentSpec;
 import io.druid.query.spec.QuerySegmentSpec;
+import jdk.nashorn.internal.ir.annotations.Ignore;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Don't use unrelated annotation

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 21]
```diff
@@ -106,7 +102,7 @@ public SegmentMetadataQuery(
       @JsonProperty("merge") Boolean merge,
       @JsonProperty("context") Map<String, Object> context,
       @JsonProperty("analysisTypes") EnumSet<AnalysisType> analysisTypes,
-      @JsonProperty("usingDefaultInterval") Boolean useDefaultInterval,
+      @Ignore @JsonProperty("usingDefaultInterval") Boolean useDefaultInterval,
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Please document that this parameter is ignored in a simple comment, also add note that it is going to be removed, and left now only for the sake of compatibility.

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 22]
```diff
@@ -121,11 +117,17 @@ public SegmentMetadataQuery(
     if (querySegmentSpec == null) {
       this.usingDefaultInterval = true;
     } else {
-      this.usingDefaultInterval = useDefaultInterval == null ? false : useDefaultInterval;
+      if (querySegmentSpec.getIntervals().size() == 1 && querySegmentSpec.getIntervals()
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Missing `.get(0)`?

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 23]
```diff
@@ -255,6 +256,14 @@ public boolean hasMinMax()
     return Druids.SegmentMetadataQueryBuilder.copy(this).toInclude(includerator).build();
   }
 
+  public SegmentMetadataQuery withFinalizedAnalysisTypes(SegmentMetadataQueryConfig config)
+  {
+    return Druids.SegmentMetadataQueryBuilder
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Don't create a new query object if analysisTypes are already non-null, just return `this`

**Comment by leventov** [comment]
@gkc2104 BTW is this valid that `SegmentMetadataQueryQueryToolChest.filterSegments()` doesn't use intervals of the query?

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java (hunk 22 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Also please add a test for this

**Comment by gkc2104** [comment]
@leventov Agreed, it is building a new interval using Default History and the max segment time.
I added a failing test that shows this as well, If you feel that the filtering is being handled incorrectly, I could update it in this PR itself

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java (hunk 22 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Please address this. Also test that if the default interval is *explicitly* specified via query builder (rather than leaving it as `null`, that will be replaced with the default interval in the `SegmentMetadataQuery`'s constructor), than isDefaultInterval is correctly set to `true`.

Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: Ok, I somehow missed `.get(0)` on the next line. The formatting is strange: suggested
```java
if (querySegmentSpec.getIntervals().size() == 1 &&
    querySegmentSpec.getIntervals().get(0).equals(DEFAULT_INTERVAL)) {
```

File: processing/src/test/java/io/druid/query/metadata/SegmentMetadataQueryTest.java [hunk 24]
```diff
@@ -1077,6 +1081,33 @@ public Interval getInterval()
     for (int i = 0; i < filteredSegments2.size(); i++) {
       Assert.assertEquals(expectedSegments2.get(i).getInterval(), filteredSegments2.get(i).getInterval());
     }
+
```
Comment (gkc2104) [inline processing/src/test/java/io/druid/query/metadata/SegmentMetadataQueryTest.java]: @leventov added the `isUsingDefaultInterval` in here along with a failing test (because of filterSegments implementation). What should I do about filterSegments though ?

**Comment by gkc2104** [comment]
Updated `filterSegments` implementation to take into consideration the query's intervals, and not return everything after DefaultHistory and nothing before even if overridden by the query.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 25]
```diff
@@ -233,18 +236,30 @@ public SegmentAnalysis apply(@Nullable SegmentAnalysis input)
...
     final T max = segments.get(segments.size() - 1);
 
     DateTime targetEnd = max.getInterval().getEnd();
-    final Interval targetInterval = new Interval(config.getDefaultHistory(), targetEnd);
+    List<Interval> intervals = query.getIntervals();
+
+    DateTime queryStartTime = JodaUtils.ETERNITY.getEnd();
+    DateTime queryEndTIme = JodaUtils.ETERNITY.getStart();
+
+    for (Interval interval : intervals) {
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Consider
```java
queryStartTime = intervals.stream().map(Interval::getEnd)
    .max(Ordering.natural()).orElseThrow(IllegalStateException::new);
```

**Comment by leventov** [comment]
@fjy restoring 0.10.1 milestone because it's an important change that restores compatibility of segment metadata queries.

@gianm @drcrallen could you please review the design again?

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java [hunk 26]
```diff
@@ -235,18 +236,27 @@ public SegmentAnalysis apply(@Nullable SegmentAnalysis input)
   @Override
   public <T extends LogicalSegment> List<T> filterSegments(SegmentMetadataQuery query, List<T> segments)
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: What's the purpose of the changes in this method? They don't seem related to the rest of the patch.

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryRunnerFactory.java [hunk 27]
```diff
@@ -87,7 +87,8 @@ public SegmentMetadataQueryRunnerFactory(
       public Sequence<SegmentAnalysis> run(QueryPlus<SegmentAnalysis> inQ, Map<String, Object> responseContext)
       {
         SegmentMetadataQuery query = (SegmentMetadataQuery) inQ.getQuery();
-        final SegmentAnalyzer analyzer = new SegmentAnalyzer(query.getAnalysisTypes());
+        SegmentMetadataQuery updatedQuery = query.withFinalizedAnalysisTypes(toolChest.getConfig());
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryRunnerFactory.java]: Might as well do this and the cast in one line to avoid accidentally referencing `query` instead of `updatedQuery`.

File: processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java [hunk 28]
```diff
@@ -106,6 +101,7 @@ public SegmentMetadataQuery(
       @JsonProperty("merge") Boolean merge,
       @JsonProperty("context") Map<String, Object> context,
       @JsonProperty("analysisTypes") EnumSet<AnalysisType> analysisTypes,
+      // useDefaultInterval will be removed, but is left for now for compatibility
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/metadata/metadata/SegmentMetadataQuery.java]: What's wrong with useDefaultInterval?

File: processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java (hunk 26 above)
Comment (gkc2104) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: Yes, these are unrelated to the changes proposed for this patch, I'll open up another PR for this.
The current implementation does not take into consideration the time range while filtering, I added an improvement to only include segments that are within the intervals mentioned by the query.

Comment (gianm) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: > The current implementation does not take into consideration the time range while filtering, I added an improvement to only include segments that are within the intervals mentioned by the query.

filterSegments doesn't need to worry about that, since the list of segments it gets has already been pruned down to whatever matches the intervals filter. In fact many implementations just `return segments;` with no changes -- check out timeseries for an example. filterSegments only has to potentially filter it down even further, if it wants (see timeBoundary for another example of that). I suppose the javadocs for filterSegments could be more clear on this.

Please also see https://github.com/druid-io/druid/pull/4259#pullrequestreview-39690653 for rationale about the current behavior of defaultInterval.

Comment (gkc2104) [inline processing/src/main/java/io/druid/query/metadata/SegmentMetadataQueryQueryToolChest.java]: I was under the impression that this method was called to prune the segments to include only the required ones from a list of all segments, didn't know it was already pruned. Now that I look at the examples you suggested I understand why it is required even if we already prune it.

Reg: `defaultInterval` - I had some failing tests that were because of `defaultInterval`, I don't remember the exact test case, but reverting to the old behavior did not break any tests (probably the change of not having null configs helped I guess), so we can keep the implementation on master 

I excluded these changes from this PR.

**Comment by gianm** [comment]
Needs one more design review.
